integration-tests/src/test/java/com/datastax/oss/driver/core/type/codec/ExtraTypeCodecsIT.java

```diff
@@ -0,0 +1,301 @@
+
+  @ClassRule
+  public static final TestRule CHAIN = RuleChain.outerRule(CCM_RULE).around(SESSION_RULE);
+
+  private enum TableField {
+    cql_text("text_value", "text"),
+    cql_int("integer_value", "int"),
+    cql_vector("vector_value", "vector<float, 3>"),
+    cql_list_of_text("list_of_text_value", "list<text>"),
+    cql_timestamp("timestamp_value", "timestamp"),
+    cql_boolean("boolean_value", "boolean"),
+    ;
+
+    final String name;
+    final String ty;
+
+    TableField(String name, String ty) {
+      this.name = name;
+      this.ty = ty;
+    }
+
+    private String definition() {
+      return String.format("%s %s", name, ty);
+    }
+  }
+
+  @BeforeClass
+  public static void setupSchema() {
+    List<String> fieldDefinitions = new ArrayList<>();
+    fieldDefinitions.add("key uuid PRIMARY KEY");
+    Stream.of(TableField.values())
+        .forEach(
+            tf -> {
+              // TODO: Move this check to BackendRequirementRule once JAVA-3069 is resolved.
+              if (tf == TableField.cql_vector
+                  && CCM_RULE.getCassandraVersion().compareTo(Version.parse("5.0")) < 0) {
+                // don't add vector type before cassandra version 5.0
+                return;
+              }
+              fieldDefinitions.add(tf.definition());
+            });
+    SESSION_RULE
+        .session()
+        .execute(
+            SimpleStatement.builder(
+                    String.format(
+                        "CREATE TABLE IF NOT EXISTS extra_type_codecs_it (%s)",
+                        String.join(", ", fieldDefinitions)))
+                .setExecutionProfile(SESSION_RULE.slowProfile())
+                .build());
+  }
+
+  private <T> void insertAndRead(TableField field, T value, TypeCodec<T> codec) {
+    try (CqlSession session =
+        (CqlSession)
+            SessionUtils.baseBuilder()
+                .addTypeCodecs(codec)
+                .addContactEndPoints(CCM_RULE.getContactPoints())
+                .withKeyspace(SESSION_RULE.keyspace())
+                .build()) {
+      // write value under new key using provided codec
+      UUID key = UUID.randomUUID();
+      session.execute(
+          SimpleStatement.builder(
+                  String.format(
+                      "INSERT INTO extra_type_codecs_it (key, %s) VALUES (?, ?)", field.name))
+              .addPositionalValue(key)
+              .addPositionalValue(value)
+              .build());
+
+      // read value using provided codec
+      assertThat(
+              session
+                  .execute(
+                      SimpleStatement.builder(
+                              String.format(
+                                  "SELECT %s FROM extra_type_codecs_it WHERE key = ?", field.name))
+                          .addPositionalValue(key)
+                          .build())
+                  .one()
+                  .get(field.name, codec))
+          .isEqualTo(value);
+    }
+  }
```

Comment:
It's not essential, but opening a new codec for each codec we want to test is probably more than we need to do here.  We can apply codecs at read time directly to a row and if we use prepared statements we can also supply a codec when we generate a BoundStatement from a PreparedStatement.  Using that approach we can do everything through a single session:

```java
  private <T> void insertAndRead(TableField field, T value, TypeCodec<T> codec) {
    CqlSession session = SESSION_RULE.session();
    PreparedStatement psi =
        session.prepare(
            String.format("INSERT INTO extra_type_codecs_it (key, %s) VALUES (?, ?)", field.name));
    // write value under new key using provided codec
    UUID key = UUID.randomUUID();
    session.execute(psi.boundStatementBuilder().setUuid(0, key).set(1, value, codec).build());

    PreparedStatement pss =
        session.prepare(
            String.format("SELECT %s FROM extra_type_codecs_it WHERE key = ?", field.name));
    // read value using provided codec
    assertThat(
            session
                .execute(pss.boundStatementBuilder().setUuid(0, key).build())
                .one()
                .get(field.name, codec))
        .isEqualTo(value);
  }
```

The above passes for me locally against C* 4.0.0.

This is not a particularly efficient way to use prepared statements, but it's probably somewhat better than having to open and close a session for each type.  As I mentioned above it's not essential (although it probably is a bit faster); there's nothing wrong with what you have now.

Reply:
Thanks. Updated `insertAndRead` to use single session with bound statements per your suggestion. If the tests pass I'll merge this one :)